docs/rfcs/041-rel-sparse-keyspace.md

```diff
@@ -0,0 +1,176 @@
+
+We will introduce a config item and an index_part record to record the current status of the migration process.
+
+* Config item `enable_reldir_v2`: controls whether the ingestion path writes the reldir info into the new keyspace.
+* `index_part.json` field `reldir_v2_status`: whether the timeline has written any key into the new reldir keyspace.
+
+If `enable_reldir_v2` is set to `true` and the timeline ingests the first key into the new reldir keyspace, it will update
+`index_part.json` to set `reldir_v2_status` to `Status::Migrating`. Even if `enable_reldir_v2` gets flipped back to
+`false` (i.e., when the pageserver restarts and such config isn't persisted), the read/write path will still
+read/write to the new keyspace to avoid data inconsistency.
+
+## Full Migration
+
+This won't be implemented in the project's first phase but might be implemented in the future. Having both v1 and
+v2 existing in the system would force us to keep the code to deserialize the old reldir key forever. To entirely deprecate this
+code path, we must ensure the timeline has no old reldir data.
+
+We can trigger a special image layer generation process at the gc-horizon. The generated image layers will cover several keyspaces:
+the old reldir key in each of the databases, and the new reldir sparse keyspace. It will remove the old reldir key while
+copying them into the corresponding keys in the sparse keyspace in the resulting image. This special process happens in
+the background during compaction. For example, assume this special process is triggered at LSN 0/180. The `create_image_layers`
+process discovers the following keys at this LSN.
+
+```
+db1/reldir_key -> (table 1, table 2, table 3)
+...db1 rel keys
+db2/reldir_key -> (table 4, table 5, table 6)
+...db2 rel keys
+sparse_reldir_db1_table1 -> deleted
+sparse_reldir_db2_table7 -> exists
+```
+
+It will generate the following keys in the new image layer:
+
+```
+db1/reldir_key -> (table 1, table 2, table 3) + migrated
+...db1 rel keys
+db2/reldir_key -> (table 4, table 5, table 6) + migrated
+...db2 rel keys
+sparse_reldir_db1_table1 -> deleted (keep it until the keyspace gets fully migrated)
+sparse_reldir_db1_table2 -> exists
+sparse_reldir_db1_table3 -> exists
+sparse_reldir_db2_table4 -> exists
+sparse_reldir_db2_table5 -> exists
+sparse_reldir_db2_table6 -> exists
+sparse_reldir_db2_table7 -> exists
+```
+
+Once we verified that no pending modifications to the old reldir exists in the delta/image layers above the gc-horizon,
+we can mark `reldir_v2_status` in the `index_part.json` to `Status::Migrated`, and the read path won't need to read from
+the old reldir anymore. Furthermore, image layer generation can drop all keys marked as `deleted` (we need to keep these
+deleted keys before the migration is fully completed in case they get created in the old reldir key but deleted in the new
+sparse keyspace).
+
+## Next Steps
+
+### Consolidate Relation Size Keys
```

Comment:
Would it make sense to track this from the beginning? Changing image format later on will be more difficult.
We'd need to update rel extend to write to the sparse keyspace.

Reply:
I plan to store it in a separate sparse keyspace so we will keep the current image format (key -> exists/deleted).